journals/apps/api/v1/search/views.py

```diff
@@ -0,0 +1,198 @@
+logger = logging.getLogger(__name__)
+
+
+class SearchViewSet(APIView):
+    """
+    View to return Journal SearchResults via RestAPI
+    """
+    permission_classes = (IsAuthenticated, )
+    # TODO: add pagination WL-1719
+
+    def get(self, request, journal_id=None):
+        """
+        Get search results for specified journal or all journals
+        for current site
+        """
+        search_results = self._search(request, journal_id)
+        if search_results:
+            serializer = SearchResultsSerializer(search_results)
+            return Response(serializer.data)
+
+        return Response(status=status.HTTP_404_NOT_FOUND)
+
+    def _search(self, request, journal_id=None):
+        '''
+        handler for search requests
+        '''
+        search_query = request.GET.get('query', None)
+        search_operator = request.GET.get('operator', 'or')
+        search_filter = request.GET.get('type', 'all')
+
+        about_pages = []
+
+        if journal_id:
+            try:
+                about_pages.append(JournalAboutPage.objects.get(journal_id=journal_id))
+            except JournalAboutPage.DoesNotExist:
+                return None
+        else:
+            # get all live journals for the requested site
+            about_pages = request.site.root_page.get_children().live().specific()
+
+        hit_list = []
+        search_meta = SearchMetaData()
+
+        if search_query:
+
+            clean_query = search_query  # TODO: do we need to do any cleansing of querystring?
+
+            #  search through each journal individual so we can check user permissions
+            for about_page in about_pages:
+
+                #  skip if the user doesn't have permissions for the specific journal
+                if not JournalAccess.user_has_access(request.user, about_page.journal):
+                    logger.warning(
+                        'User:{user} does not have access to Journal:{journal} for search'.format(
+                            user=request.user,
+                            journal=about_page.journal)
+                    )
+                    continue
+
+                #  Only include pages beneath our root page that are live and public
+                base_page_query = JournalPage.objects.live().public().descendant_of(about_page)
+
+                if search_filter == 'all':
+                    # Search pages which will yeild hits for text/HTML/RawHTML
+                    page_search_results = base_page_query.search(
+                        clean_query,
+                        operator=search_operator
+                    ).annotate_score(
+                        'score'
+                    )
+
+                    self._add_to_hit_list(hit_list, page_search_results, journal_page=True)
+                    search_meta.text_count = page_search_results.count()
+
+                if search_filter == 'all' or search_filter == 'documents':
```

Comment:
The document, video, and image searches seem very similar. I'd be nice if you could abstract them. Something like:

`filetype_search(queryset, model)` and have it return the search_results.

Reply:
good idea

Reply:
done